### DISCUSSION TITLE
Syscall proxying on Zephyr

### DISCUSSION THREAD

**Discussion by rhaberkorn** [OpenAMP/open-amp#601] [Q&A]
Hello!

I am struggling to use syscall (file IO) proxying in a Zephyr firmware running on a slave MCU (M4 core on STM32MP157C). The master will be running Linux.

Unfortunately, the [rpc_demo](apps/examples/rpc_demo/rpc_demo.c) example is not written for Zephyr. So I tried to adapt code from the [rpmsg_multi_services](https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/src/main_remote.c).

I have to call `rpmsg_rpc_init()` and pass `plaform_poll` which is unfortunately not defined on Zephyr. I believe it should be similar to this:

```c
static int platform_poll(void *priv)
{
	struct remoteproc *rproc = priv;
	int ret;

	while(1) {
		if (metal_io_read32(shm_io, 0)) {
			ret = remoteproc_get_notification(rproc,
							  RSC_NOTIFY_ID_ANY);
			if (ret)
				return ret;
			break;
		}
	}
	return 0;
}
```

Do I understand correctly, that I will have to implement something like `platform_create_proc()` as well in order to initialize `struct remoteproc`?
This leads of course into a rabbit hole, since I will now have to provide `remoteproc_ops` as well. I feel, I might be misunderstanding something fundamentally.

**Comment by arnopo** [comment]
Hello @rhaberkorn 
on zephyr we use mailbox ( or legacy ipm) for notification.
to notify the remote side we use [mailbox_notify()](https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/src/main_remote.c#L190)
To receive a message you have to register a callback in [rpmsg_create_ep()](https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/src/main_remote.c#L172)

**Reply by rhaberkorn** [reply]
Okay, you say that `ipm_send()` must be used instead of `remoteproc_get_notification()`? Actually I wasn't even sure what's the purpose of `remoteproc_get_notification()`. It's poorly documented.
What exactly the `platform_poll()` callback is supposed to do - how does it fit into the larger picture?

**Comment by tnmysh** [comment]
The platform_poll() and remoteproc_get_notifiaction() are used baremetal (mostly for zynqmp platform) and linux apps (linux userspace apps that is using OpenAMP library).

Whole apps directory needs to be refactored.

**Reply by rhaberkorn** [reply]
That doesn't really explain what they do and what I have to change for a port to Zephyr.

**Reply by tnmysh** [reply]
platform_poll() is platform specific implementation to wait till kick is received from host.
Once kick is received, remoteproc_get_notifiaction() function starts scanning vring for new buffers.

I don't know what you mean by "port to zephyr". Zephyr doesn't need to implement platform_poll(). This functionality is achieved by using mailbox or ipm driver. Please refer zephyr demo.

Zephyr demo receives messages here: https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/zephyr/rpmsg_multi_services/src/main_remote.c#L471

**Reply by rhaberkorn** [reply]
@tnmysh If I don't implement platform_poll() on Zephyr, won't these loops in rpmsg_retarget.c simply block indefinitely?
```c
while ((atomic_flag_test_and_set(&rpc->nacked))) {
        if (rpc->poll)
                rpc->poll(rpc->poll_arg);
}
```
The nacked flag is obviously cleared only in rpmsg_rpc_ept_cb(). So unless that get's called from interrupts or in separate threads, there is no chance this flag could be cleared. And indeed, my code without any poll callback appears to hang.

The rpmsg_multi_services example for Zephyr appears to use semaphores for that purpose (cf. app_rpmsg_raw()), so _cooperative_ threading appears to be involved after all.
I would suppose, that it could be worth a try to implement platform_poll() with a `k_yield()` or even `k_usleep()`.

**Comment by rhaberkorn** [comment]
btw. it even seems, I cannot currently enable the Proxy library (`WITH_PROXY`) from Zephyr. Hell, I don't even see how Zephyr eventually passes down CMake flags to OpenAMP. Apparently I will have to discuss this with the Zephyr people. Apparently they have a Discord channel for OpenAMP...

**Reply by rhaberkorn** [reply]
Theoretically, it should be possible to do a `west build -- -DWITH_PROXY=ON`, but it doesn't work. The settings gets reset/overridden in Zephyr's CMake cache.

**Reply by rhaberkorn** [reply]
This one I solved. It's a problem in Zephyr's binding which silently and mercilessly forced WITH_PROXY=OFF. Should be fixed upstream (in Zephyr).
